UIA handler/notification event handling: fetch native window handle if present in runtime ID

josephsl opened the pull request:
### Link to issue number:
Closes #16871

### Summary of the issue:
NVDA does not handle UIA notification events for elements with native window handle despite shown in runtime ID.

### Description of user facing changes
NVDA will announce UIA notifications in more apps such as Windows 11 Voice Access.

### Description of development approach
Updated UIA handler/notification event handler:

* Added type hints for event arguments.
* If UIAHandler.handler.getNearestWindowHandle method returns None (for the sender), look at the sender's (element's) runtime ID tuple, returning early if the tuple is empty.
* Pass in the just obtained native window handle to UIA NVDA object constructor.

### Testing strategy:
Manual testing (see #16862 for details):

1. Run Voice Access (Windows 11 2022 Update).
2. After confugiring the microphone, press Space on "microphone active/sleep" button.
3. Observe NVDA's UIA notificatoin event handler (should announce changed mic state).

Note that the test build must be installed or run with admin rights.

### Known issues with pull request:
In some cases runtime ID is empty, and for that case, there's nothing NVDA can do. The native window handle check is done early for performance - if UIA NVDA object constructor does not see a window handle, the constructor will call UIAHandler.handler.getNearestWindowHandle method to obtain a window handle. Also, while more UIA notifications are announced, it may need some tweaks such as making sure the window handle is indeed a native window handle, and at least runtime ID tuple check is performed in this PR; these should be dealt with on a ase-by-case basis (as issues arise).

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added support for NVDA to announce additional UIA notifications in applications, such as Windows 11 Voice Access.
  - Expanded NVDA's utilization of events for caret movement across XAML, WPF text controls, and major browsers, ensuring more consistent and responsive feedback for text cursor positioning.

- **Documentation**
  - Updated documentation to reflect the new enhancements and changes to caret movement handling and UIA notifications.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

josephsl committed:
UIAHandler: update copyright year

josephsl committed:
UIA handler/notification event handler: attempt to fetch native window handle when the element reports nothing. Re #16871.

Sometimes a UIA element may say there is no native window handle when in fact it is part of its runtime ID. Therefore fetch the runtime ID tuple if possible, logging the process along the way. If the runtime ID tuple (really an array) is empty, return early as there is nothing NVDA can do at this time.

josephsl committed:
UIA handler/notification event handling: pass in the just found native window handle to UIA object constructor. Re #16871.

Pass in the just located native window handle when instantiating UIA NVDA object to align with most event handlers 9such as focus handling).

josephsl committed:
UIA handler/notification event handling: add type hints for event arguments

josephsl committed:
Changelog: add bug fix entry for UIA notification/native window handle fetching. Re #16871

josephsl committed:
UIA handler/notification: lint

LeonarddeR commented on source/UIAHandler/__init__.py:
```diff
@@ -1037,8 +1037,39 @@ def IUIAutomationNotificationEventHandler_HandleNotificationEvent(
+					f"activityId={activityId}"
+				)
+			if any(runtimeID := sender.getRuntimeID()):
+				# Second item in runtime ID array is native window handle.
```
Can you say with certainty that this always applies?

josephsl commented:
Hi, turns out File Explorer’s folders/files view only shows three items in its runtime ID, so I’ll edit the source code comment. Thanks.

josephsl committed:
UIA handler/notification: clarify runtime ID tuple condition. Re #16871.

Comment from Leonard de Ruijter: make sure we're dealing with the correct UIA runtime ID representatio (the one with native window handle has four elements; sometimes tuple item count can vary, including an empty tuple).

LeonarddeR commented:
@coderabbitai is there anything you can say about the assumption that the window handle is part of an UIA runtime id?

LeonarddeR commented:
I know that at least accesskit creates a runtime identifier based on its internal node ID. Therefore I don't think the current approach is feasible.
Why is NVDA unable to report notifications for objects without a window handle?

josephsl commented:
Hi,

At least Code Rabbit gave some answers, but that's only part of the picture. It is correct that window handles are needed to identify UIA elements. More importantly, window handles are needed to construct a more accurate view of a UIA element, as without this, UIAHandler.handler.getNearestWindowHandle can fail to notice elements due to COM error while executing walkers. Also, creation of UIA NVDA object ensures consistency with IAccessible/IA2 objects and allows other components such as app modules and default object implementation to handle events.

The heart of the issue is inaccurate UIA element (sender) information in apps such as Voice Access when they raise notification event. The fact that another assistive technology (Narrator) must be active to see correct sender information is concerning. While I agree that retrieving window handle through runtime ID is a fragile approach, it is one sure way to identify elements (window handle check was also performed via dev info (NVDA+F1) command).

Thanks.

LeonarddeR commented:
> The heart of the issue is inaccurate UIA element (sender) information in apps such as Voice Access when they raise notification event.

This makes me think about how we handle invalid window handles with WDAG. May be that approach should be generalized instead?

> The fact that another assistive technology (Narrator) must be active to see correct sender information is concerning.
Could you elaborate?

josephsl commented:
Hi,

Generalizing WDAG approach: perhaps we could recycle that approach but to cover background events as well (WDAG is deprecated). For Voice Access, the most effective use case is dictation while not focused on Voice Access interface itself.

Comment on sender info with Narrator active: when you run Voice Access whiel using NVDA with debug logging on (with UIA included), you'll notice that NVDA will log that there is no window handle. However, when Narrator runs, notification is announced and the UIA object is created successfully; when Narrator exits, NVDA reverts to the inablity to obtain window handle.

Thanks.

seanbudd commented on source/UIAHandler/__init__.py:
```diff
@@ -1037,8 +1037,41 @@ def IUIAutomationNotificationEventHandler_HandleNotificationEvent(
+						f"displayString={displayString} "
+						f"activityId={activityId}"
+					)
+			else:
+				# No runtime ID (tuple is empty).
+				if _isDebug():
+					log.debugWarning(
+						"HandleNotificationEvent: native window handle not found in runtime ID: "
+						f"NotificationProcessing={NotificationProcessing} "
+						f"displayString={displayString} "
+						f"activityId={activityId}"
+					)
+					return
```
shouldn't this be?

```suggestion
			else:
				# No runtime ID (tuple is empty).
				if _isDebug():
					log.debugWarning(
						"HandleNotificationEvent: native window handle not found in runtime ID: "
						f"NotificationProcessing={NotificationProcessing} "
						f"displayString={displayString} "
						f"activityId={activityId}"
					)
				return
```

josephsl commented:
Hi,

Upon further investigation, while handling notification events, it turns out elements raising this event become "native" while Narrator is running, "non-native" otherwise. Therefore, this issue may require editing native UIA element fetcher (UIA handler) to tell NVDA to treat some elements as native even though it is not seen as such at first. While editing notification event handler could be a useful change, I think the issue lies deeper (this is why I'm not taking Sean's suggested edits yet).

Thanks.

LeonarddeR commented:
May be it makes sense to contact Microsoft about this?

josephsl commented:
Hi,

Sure.

I have narrowed the issue to UIAHandler.handler.getNearestWindowHandle - the method that says yes or no about native UIA element depends on nearest window handle getter. In short, tree walker and window handle caching succeeds or fails in Voice Access while Narrator is enabled and disabled, respectively.

There is a related issue that also deals with UIA notification: input language switcher. For that one, runtiem ID is undefined (empty). Given that language switch can happen everywhere, I think we should revisit the decision to limit notification handling to focused app (I think our justification back then (annoyance) does not hold well); that's for another issue.

Thanks.

michaelDCurran commented:
I would be uncomfortable accepting this pr as there is no clear evidence that the second value is always a native window handle. Clearly there is something special going on between narrator and VoiceAccess. This really needs to be talked about with Microsoft, specifically the VoiceAccess team. I am happy to reach out to them through my other Microsoft contacts and get them connected with you @josephsl   if that would help?

josephsl commented:
Hi,

I came to the same conclusion. At this point, there is no direction, and turns out the PR won't work with portable NVDA anyway. It could be a good use case for add-ons but not up to par with NVDA Core code quality. I think we could also raise a point about File Explorer/shell/input switch notification event as this is perhaps the weirdest element we came across (no runtime ID at all).

When reaching out to Microsoft, please CC my university address (joseph.lee-2@colorado.edu) as I want to use that conversation as a starting point for possible research on communication between different organizations on accessibility and usability.

Thanks.

michaelDCurran commented:
Closing based on above comments. Hopefully it can be addressed in VoiceAccess itself.
